ledger/sandbox/src/main/scala/com/digitalasset/platform/index/PostgresIndexer.scala

```diff
@@ -0,0 +1,263 @@
+import scalaz.syntax.tag._
+
+import scala.concurrent.duration._
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.util.control.NonFatal
+import scala.util.{Failure, Success}
+
+object PostgresIndexer {
+  private val logger = LoggerFactory.getLogger(classOf[PostgresIndexer])
+  private[index] val asyncTolerance = 30.seconds
+
+  def create(readService: ReadService, jdbcUrl: String): Future[PostgresIndexer] = {
+    val actorSystem = ActorSystem("postgres-indexer")
+    val materializer: ActorMaterializer = ActorMaterializer()(actorSystem)
+    val metricsManager = MetricsManager(false)
+
+    implicit val ec: ExecutionContext = DEC
+
+    val ledgerInit = readService
+      .getLedgerInitialConditions()
+      .runWith(Sink.head)(materializer)
+
+    val ledgerDao: LedgerDao = initializeDao(jdbcUrl, metricsManager)
+    for {
+      LedgerInitialConditions(ledgerIdString, _, _) <- ledgerInit
+      ledgerId = domain.LedgerId(ledgerIdString)
+      _ <- initializeLedger(ledgerId, ledgerDao)
+      ledgerEnd <- ledgerDao.lookupLedgerEnd()
+      externalOffset <- ledgerDao.lookupExternalLedgerEnd()
+    } yield {
+      new PostgresIndexer(ledgerEnd, externalOffset, ledgerDao)(materializer) {
+        override def close(): Unit = {
+          super.close()
+          materializer.shutdown()
+          Await.result(actorSystem.terminate(), asyncTolerance)
+          metricsManager.close()
+        }
+      }
+    }
+  }
+
+  private def initializeDao(jdbcUrl: String, mm: MetricsManager) = {
+    val dbDispatcher = DbDispatcher(jdbcUrl, noOfShortLivedConnections, noOfStreamingConnections)
+    val ledgerDao = LedgerDao.metered(
+      PostgresLedgerDao(
+        dbDispatcher,
+        ContractSerializer,
+        TransactionSerializer,
+        ValueSerializer,
+        KeyHasher))(mm)
+    ledgerDao
+  }
+
+  private def ledgerFound(foundLedgerId: LedgerId) = {
+    logger.info(s"Found existing ledger with id: ${foundLedgerId.unwrap}")
+    Future.successful(foundLedgerId)
+  }
+
+  private def initializeLedger(ledgerId: domain.LedgerId, ledgerDao: LedgerDao) = {
+    ledgerDao
+      .lookupLedgerId()
+      .flatMap {
+        case Some(foundLedgerId) if foundLedgerId == ledgerId =>
+          ledgerFound(foundLedgerId)
+
+        case Some(foundLedgerId) =>
+          val errorMsg =
+            s"Ledger id mismatch. Ledger id given ('$ledgerId') is not equal to the existing one ('$foundLedgerId')!"
+          logger.error(errorMsg)
+          Future.failed(new IllegalArgumentException(errorMsg))
+
+        case None =>
+          logger.info(s"Initializing ledger with id: ${ledgerId.unwrap}")
+          ledgerDao.initializeLedger(ledgerId, 0).map(_ => ledgerId)(DEC)
+      }(DEC)
+  }
+}
+
+class PostgresIndexer(initialOffset: Long, beginAfter: Option[LedgerString], ledgerDao: LedgerDao)(
+    implicit mat: Materializer)
+    extends Indexer
+    with AutoCloseable {
+
+  @volatile
+  private var headRef = initialOffset
```

Comment:
That's not quite true, because we don't increment the offset for all events we receive from the read service, so we would be producing noncontiguous offsets which would be fine I guess?

Reply:
it depends on when you are merging the offset generator in. If you do it right before you need to use them you're fine.